Add destroy pool.

jnory opened the pull request:
Hi,

I recognize that current `MemoryPool` implementation does not have features to free GPU memory, which is possibly a problem for server side applications.

This small PR gives a feature to release GPU memory forcibly.
We can now `chainer.cuda.pool.destroy()` after each jobs.

okuta commented:
Please write test.

okuta commented:
Hou about use `free_all_free`, and change method name `destroy` to `free_all_free`?

jnory commented:
Thank you for your comments. I will try them ASAP.

jnory commented:
@okuta I've added tests but some tests other than my changes (e.g. test_reversed_inner (test_product.TestProduct)) are now failing. I'm under investigation.

By the way, I have a question: Why `TestMemoryPool.pool` is instance of `SingleDeviceMemoryPool`? https://github.com/pfnet/chainer/blob/ad7a68fe5c549b477ef13003c836ae782e253472/tests/cupy_tests/cuda_tests/test_memory.py#L143

jnory commented:
Oops, I downgrade numpy to 1.9.3, all tests has passed.

okuta commented:
https://github.com/pfnet/chainer/blob/ad7a68fe5c549b477ef13003c836ae782e253472/tests/cupy_tests/cuda_tests/test_memory.py#L143

jnory commented:
I fixed it together with original PR.

okuta commented on cupy/cuda/memory.pyx:
```diff
@@ -341,6 +341,9 @@ cdef class SingleDeviceMemoryPool:
         self._free = collections.defaultdict(list)
 
+    cpdef n_free(self):
+        return sum([len(free) for _, free in self._free.iteritems()])
```
How about above code?
```python
Py_ssize_t sum
for v in self._free.itervalues():
    sum += len(v)
return sum
```

jnory replied:
Thank you! I've fixed here.

okuta commented on cupy/cuda/memory.pyx:
```diff
@@ -391,3 +394,21 @@ cdef class MemoryPool(object):
+        if dev in self._pools:
+            self._pools[dev].free_all_free()
+
+    cpdef n_free(self):
```
I think more meaningful name is good. For example, `n_free_blocks`.

jnory replied:
Ok, I've changed the name from `n_free` to `n_free_blocks`.

jnory commented:
Sorry for the late reply. I will fix them today.

jnory commented:
The code is now updated. Thank @okuta for the review.

okuta commented on cupy/cuda/memory.pyx:
```diff
@@ -341,6 +341,12 @@ cdef class SingleDeviceMemoryPool:
 
+    cpdef n_free_blocks(self):
+        cdef Py_ssize_t n = 0
+        for v in self._free.itervalues():
```
Sorry, I made a mistake.
Please use `six.itervalues`, because `itervalues` was removed in Python3.

jnory replied:
Oops, thank you. I've just fixed here now.

okuta commented:
LGTM except comment!

okuta commented on cupy/cuda/memory.pyx:
```diff
@@ -391,3 +398,21 @@ cdef class MemoryPool(object):
         return self._pools[dev].malloc(size)
+
+    cpdef free_all_free(self):
+        """release unused memory pool.
```
How about `"""Release free blocks in memory pool.""""`?

okuta commented on cupy/cuda/memory.pyx:
```diff
@@ -391,3 +398,21 @@ cdef class MemoryPool(object):
+
+    cpdef n_free_blocks(self):
+        """
+        :return:
```
Please fix style of document.

okuta commented on cupy/cuda/memory.pyx:
```diff
@@ -391,3 +398,21 @@ cdef class MemoryPool(object):
+    cpdef n_free_blocks(self):
+        """
+        :return:
+        the total number of unused memory pool.
```
How about `int: The number of free blocks.`?

okuta commented on cupy/cuda/memory.pyx:
```diff
@@ -391,3 +398,21 @@ cdef class MemoryPool(object):
+            self._pools[dev].free_all_free()
+
+    cpdef n_free_blocks(self):
+        """
```
Please write one-line document.

okuta commented on tests/cupy_tests/cuda_tests/test_memory.py:
```diff
@@ -140,14 +140,30 @@ def test_free_all_free(self):
-        mem = self.pool.malloc(0).mem
-        self.assertIsInstance(mem, memory.Memory)
-        self.assertNotIsInstance(mem, memory.PooledMemory)
+        from cupy.cuda import Device
```
Please import at the head of source.
And, don't import objects, only modules.

We use PEP8 and a part of OpenStack Style Guidelines related to general coding style as our basic style guidelines.
see http://docs.openstack.org/developer/hacking/#imports

okuta commented on cupy/cuda/memory.pyx:
```diff
@@ -6,6 +6,7 @@ from cupy.cuda import runtime
 
 from cupy.cuda cimport device
 from cupy.cuda cimport runtime
+import six
```
Please move import to L4(and create new import group), because six is a third party library.
see https://www.python.org/dev/peps/pep-0008/#imports

jnory commented:
Thank you for the comments. I will fix them today.

jnory commented:
The codes are now updated again.

okuta commented on chainer/cuda.py:
```diff
@@ -123,7 +123,8 @@ def __ne__(self, other):
 # ------------------------------------------------------------------------------
 if available:
-    cuda.set_allocator(cuda.MemoryPool().malloc)
+    pool = cuda.MemoryPool()
```
How about using more specific name (e.g. memory_pool)?

okuta commented on cupy/cuda/memory.pyx:
```diff
@@ -391,3 +398,21 @@ cdef class MemoryPool(object):
+        """Count the total number of free blocks.
+
+        Returns:
+        int: The total number of free blocks.
```
Need indent.

okuta commented on cupy/cuda/memory.pyx:
```diff
@@ -391,3 +398,21 @@ cdef class MemoryPool(object):
+    cpdef free_all_free(self):
+        """Release free blocks."""
+        dev = device.get_device_id()
+        if dev in self._pools:
```
The type of `self._pools` is defaultdict.
Is this check is necessary?

okuta commented on cupy/cuda/memory.pyx:
```diff
@@ -391,3 +398,21 @@ cdef class MemoryPool(object):
+        int: The total number of free blocks.
+        """
+        dev = device.get_device_id()
+        if dev in self._pools:
```
The type of `self._pools` is defaultdict.
Is this check is necessary?

okuta commented on cupy/cuda/memory.pyx:
```diff
@@ -1,6 +1,7 @@
 import collections
 import ctypes
 import weakref
+import six
```
Please add a blank line before this line.
`coolections`, `ctypes` and `weakref` is standard library.
`six` is related third party library.

jnory commented:
Thank you for the comments. I've just updated codes.
In order to confirm that  we can call `free_all_free()` or `n_free_blocks()` without calling `malloc()`, I've added more tests.

okuta commented:
Thank you for many fixing!
Added test is good.

okuta commented:
LGTM!

jnory commented:
Thank you for many reviews!
